torch_xla/csrc/runtime/stablehlo_composite_helper.cc

```diff
@@ -0,0 +1,404 @@
+      mlir::OpBuilder& builder,
+      const std::unordered_map<std::string, json>& json_map) {
+    llvm::SmallVector<mlir::NamedAttribute> named_attrs;
+    for (auto& [key, j] : json_map) {
+      switch (j.type()) {
+        case json::value_t::number_integer:
+        case json::value_t::number_unsigned:
+          named_attrs.push_back(
+              {builder.getStringAttr(key),
+               builder.getI64IntegerAttr(j.template get<int64_t>())});
+          break;
+        case json::value_t::number_float:
+          named_attrs.push_back(
+              {builder.getStringAttr(key),
+               builder.getF32FloatAttr(j.template get<float>())});
+          break;
+        case json::value_t::boolean:
+          named_attrs.push_back({builder.getStringAttr(key),
+                                 builder.getBoolAttr(j.template get<bool>())});
+          break;
+        case json::value_t::string:
+          named_attrs.push_back(
+              {builder.getStringAttr(key),
+               builder.getStringAttr(j.template get<std::string>())});
+          break;
+        default:
+          // Ignored unrecognizable attr json
+          break;
+      }
+    }
+    return builder.getDictionaryAttr(named_attrs);
+  }
+
+  void BuildStableHLOCompositeOp(
+      mlir::Operation* op,
+      const llvm::DenseMap<const mlir::Operation*, size_t>& op_line_num) {
+    mlir::ModuleOp module_op = getOperation();
+    mlir::MLIRContext* context = &getContext();
+    mlir::OpBuilder builder(context);
+
+    std::unique_ptr<BoundaryMetadata> metadata = GetBoundaryMetadata(op);
+    if (metadata == nullptr || metadata->is_input) {
+      return;
+    }
+    const auto& output_metadata = *metadata;
+
+    llvm::SetVector<mlir::Operation*> scope_ops_setvec;
+    llvm::SetVector<std::pair<mlir::Value, int64_t>> arg_pos_setvec;
+    llvm::SmallVector<mlir::Operation*> processing({op});
+
+    // Reverse graph traversal: from boundary output op to boundary input op,
+    // global function arg, or stablehlo constant.
+    while (!processing.empty()) {
+      mlir::Operation* curr_op = processing.back();
+      processing.pop_back();
+      if (scope_ops_setvec.contains(curr_op)) {
+        continue;
+      }
+
+      if (auto curr_metadata_ptr = GetBoundaryMetadata(curr_op);
+          curr_metadata_ptr != nullptr) {
+        const auto& curr_metadata = *curr_metadata_ptr;
+        if (curr_metadata.is_input &&
+            curr_metadata.boundary_id() == output_metadata.boundary_id()) {
+          // Terminal condition: boundary input op.
+          arg_pos_setvec.insert({curr_op->getResult(0).dyn_cast<mlir::Value>(),
+                                 curr_metadata.pos});
+          continue;
+        }
+      }
+
+      scope_ops_setvec.insert(curr_op);
+      for (mlir::Value value : curr_op->getOperands()) {
+        mlir::Operation* def_op = value.getDefiningOp();
+        if (def_op == nullptr) {
+          // Terminal condition: Global function arg
```

Comment:
Should this ever happen? This feels like an error since everything should eventually come through a `mark_begin`?

Reply:
Because the StableHLO is lowered from functionalized FX graph, there can be some constants lifted to be MLIR inputs.

Reply:
Ah, yeah. Thinking through it probably better to allow input arguments instead of require that all constants be marked. The alternative IIUC is:

```py
def f():
  x = 3
  x = mark_begin(x, "a")
#  -functionalize->
def f(x):
  x = mark_begin(x, "a")
```

But there are going to be constants used everywhere in a graph and to mark them all would make for some ugly code. Feel free to resolve this comment if my understanding is correct.